server/http_tunnel.go

```diff
@@ -0,0 +1,77 @@
+package server
+
+import (
+	"crypto/tls"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"time"
+)
+
+func handleTunneling(w http.ResponseWriter, r *http.Request) {
+	log.Printf("HTTP CONNECT: opening connection to %s\n", r.Host)
+	destConn, err := net.DialTimeout("tcp", r.Host, 10*time.Second)
+	if err != nil {
+		log.Printf("HTTP CONNECT: failed to open connection to %s\n", r.Host)
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		http.Error(w, "Hijacking not supported", http.StatusInternalServerError)
+		return
+	}
+	clientConn, _, err := hijacker.Hijack()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+	}
+	go transfer(destConn, clientConn)
```

Comment:
I don't think this will work as expected. These two goroutines will stop working since `handleTunneling()` already return, which means this function would wait until one of this has errors. And also, if one of the goroutines returns itself because io.Copy() returns an error, the other one should be notified as well and be closed right away.

Reply:
Not sure what you mean.  When handleTunneling() returns, I expect the two goroutines to continue to run until either EOF or Error is encountered.  This will allow all of the data in the tunnel to get copied.  When an error occurs, I expect it to affect both goroutines, since they're acting on the same underlying connections.

Reply:
Here is an example: https://play.golang.org/p/79ZLfm2-CbH

As you can see, main() returns and it won't wait until goroutine finish the print out.

Reply:
Our main goroutine is waiting on a channel read that never happens: https://github.com/PelionIoT/edge-proxy/blob/1e03559a81fe974e346598b5025862337c2c9081/cmd/edge-proxy/main.go#L193

Reply:
Yes, so this will keep goroutines running and blocks the return of main func.

Reply:
This means that `handleTunneling()` returning doesn't cause the goroutines to stop working, correct?  Is this conversation resolved?